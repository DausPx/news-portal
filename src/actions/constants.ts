export const actions = {
    DEFAULT_ACTION: "default_action",
    ADD_ARTICLES: "add_articles",
    ADD_PAGE: "add_page",
    FETCHED_ALL: "fetched_all",
  }