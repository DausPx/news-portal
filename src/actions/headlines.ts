import { actions } from "./constants";
import { IArticle } from "../reducers/topHeadlines";

export const addArticles = (articles: IArticle[]) => {
  return { type: actions.ADD_ARTICLES, payload: articles };
};

export const addPage = () => {
  return { type: actions.ADD_PAGE };
};

export const fetchedAll = ()=>{
  return { type: actions.FETCHED_ALL };
}
