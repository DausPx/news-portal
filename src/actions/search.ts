import { actions } from "./constants";
import { Dispatch } from "redux";
import API from "../api/api";
import { IArticle } from "../reducers/topHeadlines";

export const getSearchArticles = (q: string, sortBy: string) => {
  return (dispatch: Dispatch) => {
    dispatch({ type: actions.LOADING_TRUE });
    return API.get("everything", {
      params: {
        q: encodeURI(q),
        sortBy: sortBy,
        page: 1,
        pageSize: 20,
      },
    })
      .then((result) => {
        dispatch({
          type: actions.ADD_SEARCH_ARTICLES,
          payload: result.data.articles,
        });
      })
      .catch((err) => {
        console.log(err);
        dispatch({ type: actions.LOADING_FALSE });
      });
  };
};

export const LoadNextPage = (q: string, sortBy: string, page: number) => {
  return (dispatch: Dispatch) => {
    dispatch({ type: actions.LOADING_TRUE });
    return API.get("everything", {
      params: {
        q: encodeURI(q),
        sortBy: sortBy,
        page: page + 1,
        pageSize: 20,
      },
    })
      .then((result) => {
        dispatch({
          type: actions.ADD_SEARCH_PAGE,
          payload: result.data.articles,
        });

        if (result.data.articles.length === 0) {
          dispatch({ type: actions.FETCHED_ALL_SEARCH });
        }
      })
      .catch((err) => {
        console.log(err);
        dispatch({ type: actions.LOADING_FALSE });
      });
  };
};

export const setSearchQuery = (query: string) => {
  return { type: actions.SET_SEARCH_QUERY, payload: query };
};

export const setActiveArticle = (article: IArticle) =>{
  return {type: actions.SET_ACTIVE_ARTICLE, payload: article}
}

export const setSortBy = (sortBy:string)=>{
  return {type: actions.SET_SORT_BY, payload:sortBy}
}