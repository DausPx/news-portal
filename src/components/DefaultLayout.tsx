import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { NavLink, useHistory } from "react-router-dom";
import { setSearchQuery } from "../actions/search";

export type DefaultLayoutProps = {
  children: React.ReactNode;
};
const DefaultLayout = (props: DefaultLayoutProps): JSX.Element => {
  const [searchText, setSearchText] = useState<string>("");
  const dispatch = useDispatch();
  const history = useHistory();

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchText(e.target.value);
  };

  const onKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter" && searchText !== "") {
      dispatch(setSearchQuery(searchText));
      history.push(`/search`);
    }
  };

  const onSearch = () => {
    if (searchText !== "") {
      dispatch(setSearchQuery(searchText));
      history.push(`/search`);
    }
  };

  return (
    <div className="w-full">
      <header className="bg-blue-500 text-gray-600 body-font sticky">
        <div className="container mx-auto flex flex-wrap p-5 flex-col md:flex-row items-center">
          <nav className="flex md:w-2/5 flex-wrap items-center text-base md:ml-auto">
            <NavLink className="mr-5 font-bold hover:text-gray-900" to="/">
              Home
            </NavLink>
          </nav>
          <div className="md:w-3/5 inline-flex flex-col md:flex-row md:justify-start ml-5 md:ml-0">
            <input
              placeholder="type to search"
              type="text"
              value={searchText}
              onChange={onChange}
              onKeyDown={onKeyDown}
              className="inline-flex border-0 py-1 px-3 rounded text-base mt-4 md:mr-2 md:mt-0"
            />
            <button
              className="flex justify-center bg-gray-100 border-0 py-1 px-3 focus:outline-none hover:bg-gray-200 rounded text-base mt-4 md:mt-0"
              onClick={onSearch}
            >
              Search
            </button>
          </div>
        </div>
      </header>
      {/* <div className="flex w-full h-12 bg-green-500">
        <NavLink
          activeStyle={{
            backgroundColor: "rgba(107, 114, 128, var(--tw-bg-opacity))",
          }}
          to="/"
          className="mx-2"
        >
          Home
        </NavLink>
        <input
          placeholder="To search"
          type="text"
          value={searchText}
          onChange={onChange}
          onKeyDown={onKeyDown}
          className="w-24 h-full mx-2"
        />
        <button className="w-24 h-full mx-2 bg-gray-800" onClick={onSearch}>
          Search
        </button>
      </div> */}
      <div className="w-full">{props.children}</div>
    </div>
  );
};
export default DefaultLayout;
