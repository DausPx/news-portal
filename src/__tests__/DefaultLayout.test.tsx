import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import React, { Dispatch } from "react";
import DefaultLayout from "../components/DefaultLayout";
import * as reactRedux from "react-redux";
import { actions } from "../actions/constants";

configure({ adapter: new Adapter() });

afterEach(() => {
  jest.clearAllMocks();
});

const mockedPush = jest.fn();
jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useHistory: () => ({
    push: mockedPush,
  }),
}));

describe("snapshot testing and Unit Testing", () => {
  // const useSelectorMock = jest.spyOn(reactRedux, "useSelector");
  const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");

  it("renders with children", () => {
    const dummyDispatch = jest.fn();
    useDispatchMock.mockReturnValue(dummyDispatch);

    const wrapper = shallow(
      <DefaultLayout>
        <p className="test">Text</p>
      </DefaultLayout>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("onKeyDown is called as inteded", () => {
    const dummyDispatch = jest.fn();
    useDispatchMock.mockReturnValue(dummyDispatch);

    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    let init = "start";
    useStateSpy.mockImplementation((): [unknown, Dispatch<unknown>] => [
      init,
      setState,
    ]);

    const wrapper = shallow(
      <DefaultLayout>
        <p className="test">Text</p>
      </DefaultLayout>
    );

    wrapper.find("input").simulate("keydown", { key: "Enter" });

    expect(dummyDispatch).toBeCalledTimes(1);

    expect(mockedPush).toBeCalledTimes(1);

    expect(mockedPush).toBeCalledWith("/search");

    expect(dummyDispatch).toBeCalledWith({
      type: actions.SET_SEARCH_QUERY,
      payload: init,
    });
  });

  it("onSearch is called as inteded", () => {
    const dummyDispatch = jest.fn();
    useDispatchMock.mockReturnValue(dummyDispatch);

    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    let init = "start2";
    useStateSpy.mockImplementation((): [unknown, Dispatch<unknown>] => [
      init,
      setState,
    ]);

    const wrapper = shallow(
      <DefaultLayout>
        <p className="test">Text</p>
      </DefaultLayout>
    );

    wrapper.find("button").simulate("click");

    expect(dummyDispatch).toBeCalledTimes(1);

    expect(mockedPush).toBeCalledTimes(1);

    expect(mockedPush).toBeCalledWith("/search");

    expect(dummyDispatch).toBeCalledWith({
      type: actions.SET_SEARCH_QUERY,
      payload: init,
    });
  });

  it("onChange is called as inteded", () => {
    const dummyDispatch = jest.fn();
    useDispatchMock.mockReturnValue(dummyDispatch);

    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    let init = "start2";
    useStateSpy.mockImplementation((): [unknown, Dispatch<unknown>] => [
      init,
      setState,
    ]);

    const wrapper = shallow(
      <DefaultLayout>
        <p className="test">Text</p>
      </DefaultLayout>
    );

    wrapper.find("input").simulate("change", { target: { value: "test" } });

    expect(setState).toBeCalledWith("test");
  });
});
