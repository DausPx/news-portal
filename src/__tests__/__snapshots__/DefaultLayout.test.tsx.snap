// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing and Unit Testing renders with children 1`] = `
<div
  className="w-full"
>
  <header
    className="bg-blue-500 text-gray-600 body-font sticky"
  >
    <div
      className="container mx-auto flex flex-wrap p-5 flex-col md:flex-row items-center"
    >
      <nav
        className="flex md:w-2/5 flex-wrap items-center text-base md:ml-auto"
      >
        <NavLink
          className="mr-5 font-bold hover:text-gray-900"
          to="/"
        >
          Home
        </NavLink>
      </nav>
      <div
        className="md:w-3/5 inline-flex flex-col md:flex-row md:justify-start ml-5 md:ml-0"
      >
        <input
          className="inline-flex border-0 py-1 px-3 rounded text-base mt-4 md:mr-2 md:mt-0"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="type to search"
          type="text"
          value=""
        />
        <button
          className="flex justify-center bg-gray-100 border-0 py-1 px-3 focus:outline-none hover:bg-gray-200 rounded text-base mt-4 md:mt-0"
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </div>
  </header>
  <div
    className="w-full"
  >
    <p
      className="test"
    >
      Text
    </p>
  </div>
</div>
`;
